### PR TITLE
some experimental changes

### PR DESCRIPTION
**Pull request by sa-lee**
This PR introduces some changes for changing/loading data into a widget (with and without shiny). An example of how this works is given in `test_app/gapminder.R` and the result is two new exported functions called `vw_change_data()` (used with shiny) and `vw_load_data()` (loads data on render). 

It also refactors the vegawidget.js code based on the changes introduced by @AliciaSchep, with some general tidy ups to the changes introduced by the last PR. Happy to make further changes here or revert back to the previous style if you are not happy with the modifications on the js side.

### CODE REVIEW

File: inst/htmlwidgets/vegawidget.js [hunk 1]
```diff
@@ -15,47 +15,64 @@ HTMLWidgets.widget({
...
+            el.removeAttribute("style");
+          })
+          .catch(console.error);
       },
 
       resize: function(width, height) {
 
       },
 
+      // the view can just be an attribute of the HTMLWidgets object
```
Comment (AliciaSchep) [inline inst/htmlwidgets/vegawidget.js]: Is this comment suggesting that the getView function can probably be removed? I think that'd probably be a reasonable thing to do

**Comment by AliciaSchep** [comment]
JS changes look good!  

I think this will require some slight changes to the functions in to-image.R that don't seem to have been made in the PR yet though.  e.g. in https://github.com/sa-lee/vegawidget/blob/shiny-interaction/R/to-image.R#L88 from `result.toSVG` to `result.view.toSVG` and the same in the vw_to_png function.

**Comment by ijlyttle** [comment]
Hi @sa-lee,

Thanks! I have had a quick look, and I am encouraged by @AliciaSchep's reaction.

A couple of quick points in the mean time (it may be a day or so before I can clear my plate enough to try to wrap my head around everything).

Can you make the JS changes instead to the file `data-raw/templates/vegawidget.js` then render the file `data-raw/infrastructure.Rmd`? This is an imperfect system (I still need to clean it up) to make sure we get a fresh copy of the htmlwidgets directory any time we change something (particularly Vega versions).

As well, can you have a look at the style we are using for the R documentation? I have an (imperfect) contributing guide in `master` at `CONTRIBUTING.md`. 

I apologize in advance for focusing so much on style, particularly when I have not addressed the content at all - but I feel strongly that a consistent style will help our users and future contributors (and me!).

Regarding @AliciaSchep's comment on the changes that will have to be made to the image functions, if you can make those changes in the PR, I will be happy to make those changes right-away in `master`.

Thanks!

**Comment by sa-lee** [comment]
Thanks for the feedback @AliciaSchep and @ijlyttle! I'll get around to making these changes in a jiffy.
 
@ijlyttle no worries - I'll fix these things up and read over the contributing guide. Re the infrastructure.Rmd I'll make sure to follow that workflow (seems like a makefile might be a good way to go here in the future - if you want I can make (;P) one in a new PR)

**Comment by ijlyttle** [comment]
@sa-lee Thanks - a makefile could be very useful! I have this bad habit of improving a workflow just enough so that I am not totally annoyed by it, but not enough to where it can be truly useful 😔

In other words, your suggestions and improvement are very welcome!

**Comment by sa-lee** [comment]
I'm getting a weird error message when I try to run the `to-image.R` that aren't present in master. Any ideas why? getVegaView is still in vegawidget.js and runs if I look at the console in a browser.

```r
spec_mtcars %>% vw_to_svg()
Error in session_makeRequest(self, private, endpoint, data, params, headers) : 
  Can't find variable: getVegaView
```

**Comment by ijlyttle** [comment]
This may be a silly question, but you are building-and-reloading instead of using the devtools build?

**Comment by sa-lee** [comment]
I'm using `devtools::install()` then reloading.

**Comment by ijlyttle** [comment]
There's a problem between devtools and htmlwidgets (not only for vegawidget) - the workaround is to (unfortunately always) build & reload.

Does that help?

**Comment by sa-lee** [comment]
how weird - I've done that but its still not working :( , will have to investigate further.

**Comment by AliciaSchep** [comment]
@sa-lee I can reproduce your error with your fork's branch.  The error does not occur with the current vegawidget shiny-interaction branch

**Comment by AliciaSchep** [comment]
Another thing that I noticed with your branch @sa-lee is that the plot doesn't show up in my Rstudio console window.  But it does show up when opening up in a browser.  Weird...

**Comment by AliciaSchep** [comment]
@sa-lee I think the issue is one of browser compatibility.  The 'let' and '=>' usage in your version is not compatible with older browsers (like my rstudio console window or apparently whatever webdriver is using).  Changing back the 'let' to 'var' and '=>' to function() fixed the issue on my computer.  The rstudio console issue was actually helpful for debugging because the console told me what was causing the errors ('let').  I think with webdriver it was failing to load the widget but then moving on and trying the next step...

**Comment by ijlyttle** [comment]
Thanks for catching that @AliciaSchep!

As a side note, I have been trying to learn some JavaScript (learn more formally), and I had read about `let` being the modern, perhaps preferred, variant of `var`. In the back of my head, I was wondering why I was seeing `var` in the context of Vega. **Now it all makes sense.** 

At the risk of stating the obvious, this seems like a huge challenge for working with browser-based visualization - to make a conscious choice about what browsers we (the larger we) will support, then keeping in mind all the tricks/hacks/etc. to make that work.

**Comment by sa-lee** [comment]
Hi @AliciaSchep - nice, that makes total sense. Just out of curiosity what version of rstudio are you using? I am using 1.2 and the widgets render for me. I'll downgrade the fork to avoid modern js stuff like let and fat arrow notation. 

@ijlyttle one possibility in the future is to use a JS transpiler but for now I'll just make the changes @AliciaSchep suggested.

File: inst/htmlwidgets/vegawidget.js (hunk 1 above)
Comment (ijlyttle) [reply inst/htmlwidgets/vegawidget.js]: I agree - I think `getView` can be removed; I think the only usage is in `getVegaView`

**Comment by ijlyttle** [comment]
Hi @sa-lee (and @AliciaSchep),

I am starting to wrap my head around all of this, so I will have to ask your forgiveness (as usual) for my uninformed questions.

As discussed above, there are a few ideas that I would like to implement right away in master. I think it will make the eventual merge of this branch easier if we agree on these points-of-contact. Even better if the code will be the same so we can avoid merge conflicts. 🤞

## Points-of-contact

- `inst/templates/htmlwidgets/vegawidget.js`: `renderValue`

In the vega-embed examples (like [this](https://vega.github.io/vega-lite/tutorials/streaming.html)), the convention is to refer to the view promise as `result.view` (or `res.view`), rather than as `view`.

It seems we are adopting this same convention for the sake of consistency, which I agree with (I hope this is not a *foolish* consistency).

Here, what we call `view_promise`, could we call it instead `vega_promise`? If I understand correctly, the `vega_promise` has a `view` element.

- `inst/templates/htmlwidgets/vegawidget.js`: `getVegaView`

Continuing this thought, could this be renamed `getVegaPromise()`, and modified so that it simply returns the object's `vega_promise`?

This would (I think) let us remove the `getView()` function from the factory, per @AliciaSchep's suggestion.

- `R/to-image.R` 

Here, we would have to use `getVegaPromise` rather than `getVegaView`, then refer to `result.view` (which is more-conventional) rather than `result`.

I can try these changes out myself, and report back - but I would be very interested in your "philosophical" thoughts, thanks!

**Comment by AliciaSchep** [comment]
@ijlyttle I think another option is instead of defining a `getPromise` function, you can expose the object directly:

```js
vega_promise: vega_promise
```

instead of 

```js
getPromise: function() {
        return vega_promise;
      }
```

**Comment by ijlyttle** [comment]
@AliciaSchep 

Thanks, I'll make those changes 🙏

I need to improve my JS skills... in the meantime, I sure do appreciate you and @sa-lee helping me stay on the "right path"!

**Comment by sa-lee** [comment]
hi @ijlyttle and @AliciaSchep sorry for not responding sooner. I was in an email free zone over the weekend. I like the idea of including the `vega_promise` as part of the widget object. I think one point of confusion, is that they way I've structured getVegaView currently as that it resolves the promise and returns the view object (so it's name is accurate), so the functions in `to-image.R` do not need to be changed. If you did want to keep everything consistent than that function `getVegaView` should be modified to just return the promise itself.

**Comment by ijlyttle** [comment]
Hi @sa-lee - no worries - hope you enjoyed your weekend!

I have made some changes in master so that `getVegaView` has (1) been renamed `getVegaPromise`, and (2) returns the promise object rather than the view object.

I had done this because the vega-embed documentation uses the convention:

```js
.then(function(result){
  // do something to result.view
});
```

And I thought that it might reduce (long-term) confusion to follow the vega-embed convention.

I have already changed the image functions in master to conform to this, and they work for me, so you would have a template there. 

If I get a little time this afternoon, I will make the change that Alicia suggested.

Would this work for you?

**Comment by ijlyttle** [comment]
I tried to expose just the `vega_promise` object in the factory, but I could not get it to work - it would return as `null` - the initialized value.

Of course I could be doing something wrong (as is my habit), but I wonder if exposing it as a function somehow subjects it to lazy evaluation. 

I did a little googling and found that JS has a `get` syntax https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get

In the `vegalite-3` branch, I put tried out some code like this ([link](https://github.com/vegawidget/vegawidget/blob/e0999cb26a1238643322549e8ad846a593a740c7/data-raw/templates/vegawidget.js#L37)):

```js
get vegaPromise() {
  return vega_promise;
}
```

then referred to it using `foo.vegaPromise`.

It *seems* to work, how about this?

**Comment by AliciaSchep** [comment]
a delayed thanks @ijlyttle, did not know about 'get' in JS before!  sounds like a good solution

**Comment by AliciaSchep** [comment]
@sa-lee and @ijlyttle  What do y'all think about merging this now?  (into shiny-interaction branch) It seems like a good additional set of functionality, and if there are proposed refactors or API changes those could be new issues/PRs targeting the shiny-interaction branch.  We could then also go ahead and merge/reconcile changes made to master with the shiny interaction branch.

**Comment by ijlyttle** [comment]
This is OK by me - I'll wait for any objection from @sa-lee...

**Comment by sa-lee** [comment]
No objections from me either.

**Comment by ijlyttle** [comment]
Let it be done!

**Pull request merged**